### PR TITLE
refactor(ChangelogFormat): refactor get_metadata_from_file for better readability and type

### PR DESCRIPTION
**Pull request by bearomorphism**
… readability and better type

## Checklist

- [ ] I have read the [contributing guidelines](https://commitizen-tools.github.io/commitizen/contributing/)

### Code Changes

- [ ] Add test cases to all the changes you introduce
- [ ] Run `poetry all` locally to ensure this change passes linter check and tests
- [ ] Manually test the changes:
  - [ ] Verify the feature/bug fix works as expected in real-world scenarios
  - [ ] Test edge cases and error conditions
  - [ ] Ensure backward compatibility is maintained
  - [ ] Document any manual testing steps performed
- [ ] Update the documentation for the changes

### Documentation Changes

- [ ] Run `poetry doc` locally to ensure the documentation pages renders correctly
- [ ] Check and fix any broken links (internal or external) in the documentation

> When running `poetry doc`, any broken internal documentation links will be reported in the console output like this:
>
> ```text
> INFO    -  Doc file 'config.md' contains a link 'commands/bump.md#-post_bump_hooks', but the doc 'commands/bump.md' does not contain an anchor '#-post_bump_hooks'.
> ```

### CODE REVIEW

File: commitizen/changelog_formats/base.py [hunk 1]
```diff
@@ -40,34 +40,40 @@ def get_metadata(self, filepath: str) -> Metadata:
...
         unreleased_level: int | None = None
-        for index, line in enumerate(file):
-            line = line.strip().lower()
 
-            unreleased: int | None = None
-            if "unreleased" in line:
-                unreleased = self.parse_title_level(line)
+        lines = file.readlines()
+        for index, line in enumerate(line.strip().lower() for line in lines):
+            parsed_unreleased_level = self.parse_title_level(line)
```
Comment (bearomorphism) [inline commitizen/changelog_formats/base.py:49]: I think we can assume that `self.parse_title_level` is always a pure function? Then we don't have to call it twice here.

File: commitizen/changelog_formats/base.py [hunk 2]
```diff
@@ -40,34 +40,40 @@ def get_metadata(self, filepath: str) -> Metadata:
...
-                meta.latest_version = parsed.version
-                meta.latest_version_tag = parsed.tag
-                meta.latest_version_position = index
+            if parsed_version_tag := self.parse_version_from_title(line):
+                out_metadata.latest_version = parsed_version_tag.version
+                out_metadata.latest_version_tag = parsed_version_tag.tag
+                out_metadata.latest_version_position = index
                 break  # there's no need for more info
-        if meta.unreleased_start is not None and meta.unreleased_end is None:
-            meta.unreleased_end = index
```
Comment (bearomorphism) [inline commitizen/changelog_formats/base.py:71]: This `index` is unbounded if the file is empty, although this never happens for all code paths

File: commitizen/changelog_formats/base.py (hunk 1 above)
Comment (Lee-W) [reply commitizen/changelog_formats/base.py:49]: If we want to make this assumption, we'll need to make this a static method instead, which makes sense but could be seen as a breaking change.

File: commitizen/changelog_formats/base.py [hunk 3]
```diff
@@ -40,34 +40,40 @@ def get_metadata(self, filepath: str) -> Metadata:
...
+    def get_metadata_from_file(self, file: IO[str]) -> Metadata:
+        out_metadata = Metadata()
         unreleased_level: int | None = None
-        for index, line in enumerate(file):
-            line = line.strip().lower()
 
-            unreleased: int | None = None
-            if "unreleased" in line:
-                unreleased = self.parse_title_level(line)
+        lines = [line.strip().lower() for line in file.readlines()]
```
Comment (Lee-W) [inline commitizen/changelog_formats/base.py:47]: loop through it twice, probably not worth it?

File: commitizen/changelog_formats/base.py (hunk 1 above)
Comment (Lee-W) [reply commitizen/changelog_formats/base.py:49]: but making that a static method is a good idea
